File: MatchstickstoSquare3.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
public:
    bool makesquare(vector<int>& nums) {
        if(nums.size() <= 3) return false;
        int sum = 0;
        for(int num : nums) sum += num;
        if(sum % 4) return false;
        int len = sum / 4;
        int n = nums.size();
        int all = (1 << n) - 1;
        vector<int> masks;
        vector<bool> validHalf(1 << n, false);
        
        for(int i = 1; i <= all; i++) {
        	int curSum = 0;
        	for(int j = 0; j < 15; j++) {
        		if((1 << j) & i) curSum += nums[j];
			}
			
			if(curSum == len) {
				for(int mask : masks) {
					if(mask & i) continue;
					int half = mask | i;
					validHalf[half] = true;
					if(validHalf[all ^ half]) return true;
				}
				masks.push_back(i);
			}
		}
		return false;
    }
};

int main() {
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->makesquare(nums) ? cout<<"true" : cout<<"false";
	
	return 0;
}

File: MatchstickstoSquare3.py
class Solution:
    def makesquare(self, nums) -> bool:
        if len(nums) <= 3:
            return False
        total = sum(nums)
        if total % 4 != 0:
            return False
        target = total // 4
        n = len(nums)
        length = (1 << n) - 1

        masks = []
        validHalf = set()
        for i in range(1, length + 1, 1):
            curSum = 0
            for j in range(n):
                if (1 << j) & i:
                    curSum += nums[j]
            if curSum == target:
                for mask in masks:
                    if i & mask:
                        continue
                    half = i | mask
                    validHalf.add(half)
                    if length ^ half in validHalf:
                        return True
                masks.append(i)
        return False


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.makesquare(nums))
